crates/adapters/src/integrated/delta_table/test.rs

```diff
@@ -0,0 +1,846 @@
+use env_logger::Env;
+use log::{debug, info, trace};
+use parquet::file::reader::Length;
+use pipeline_types::config::PipelineConfig;
+use pipeline_types::program_schema::{Field, Relation};
+use pipeline_types::serde_with_context::serialize::SerializeWithContextWrapper;
+use pipeline_types::serde_with_context::{
+    DateFormat, DeserializeWithContext, SerializeWithContext, SqlSerdeConfig, TimeFormat,
+    TimestampFormat,
+};
+use pipeline_types::transport::delta_table::DeltaTableIngestMode;
+use proptest::collection::vec;
+use proptest::prelude::{Arbitrary, ProptestConfig, Strategy};
+use proptest::proptest;
+use proptest::strategy::ValueTree;
+use proptest::test_runner::TestRunner;
+use serde_arrow::schema::SerdeArrowSchema;
+use std::collections::HashMap;
+use std::ffi::OsStr;
+use std::fs::File;
+use std::io::{Read, Write};
+use std::mem::forget;
+use std::os::unix::ffi::OsStrExt;
+use std::path::Path;
+use std::sync::Arc;
+use std::time::{Duration, Instant};
+use tempfile::{tempdir, NamedTempFile, TempDir};
+use tokio::time::sleep;
+use uuid::Uuid;
+
+fn delta_output_serde_config() -> SqlSerdeConfig {
+    SqlSerdeConfig::default()
+        .with_date_format(DateFormat::String("%Y-%m-%d"))
+        // DeltaLake only supports microsecond-based timestamp encoding, so we just
+        // hardwire that for now.  See also `format/parquet/mod.rs`.
+        .with_timestamp_format(TimestampFormat::MicrosSinceEpoch)
+}
+
+/// Read a snapshot of a delta table to a temporary JSON file.
+fn delta_table_snapshot_to_json<T>(
+    table_uri: &str,
+    config: &HashMap<String, String>,
+) -> NamedTempFile
+where
+    T: DBData
+        + SerializeWithContext<SqlSerdeConfig>
+        + for<'de> DeserializeWithContext<'de, SqlSerdeConfig>
+        + Sync,
+{
+    let start = Instant::now();
+    let json_file = NamedTempFile::new().unwrap();
+    println!(
+        "delta_table_snapshot_to_json: writing output to {}",
+        json_file.path().display()
+    );
+
+    let mut config = config.clone();
+    config.insert("mode".to_string(), "snapshot".to_string());
+
+    let input_pipeline = delta_table_input_pipeline::<T>(
+        table_uri,
+        &config,
+        &json_file.path().display().to_string(),
+    );
+    input_pipeline.start();
+    wait(|| input_pipeline.status().pipeline_complete(), 400_000).expect("timeout");
+    input_pipeline.stop().unwrap();
+
+    info!("Read delta snapshot in {:?}", start.elapsed());
+
+    json_file
+}
+
+/// Wait until `table` contains exactly `expected_count` records.
+async fn wait_for_count_records(
+    table: &mut Arc<DeltaTable>,
+    expected_count: usize,
+    datafusion: &SessionContext,
+) {
+    let start = Instant::now();
+    loop {
+        // select count() output_table == len().
+        Arc::get_mut(table)
+            .unwrap()
+            .update_incremental(None)
+            .await
+            .unwrap();
+
+        let count = datafusion
```

Comment:
I hope this is efficient

Reply:
It should be relatively efficient. It should only look at the metadata of each file.